File: Pset 6 (Python)/readability.py
from cs50 import get_string


def main():
    # prompt the user for text
    text = get_string("Text: ")

    # count the number of letters, words, and sentences in the text
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # compute the Coleman-Liau index
    L = letters / words * 100
    S = sentences / words * 100
    index = 0.0588 * L - 0.296 * S - 15.8
    # round to the nearest integer
    grade = round(index)

    # print the grade level
    if grade < 1:
        print("Before Grade 1")
    elif grade > 16:
        print("Grade 16+")
    else:
        print(f"Grade {grade}")


def count_letters(text):
    # for each alphabetical char in text, add 1
    letters = sum(1 for char in text if char.isalpha())
    return letters


def count_words(text):
    # split() splits a string into a list of words
    words = len(text.split())
    return words


def count_sentences(text):
    # for any of . ! or ? in text, add 1
    sentences = sum(1 for char in text if char in ".!?")
    return sentences


main()

File: Pset 2 (Arrays)/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Prompt the user for some text
    string text = get_string("Text: ");

    // Count the number of letters, words, and sentences in the text
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Compute the Coleman-Liau index
    float L = (float) letters / (float) words * 100;
    float S = (float) sentences / (float) words * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;
    // Round to the nearest integer
    int grade = round(index);

    // Print the grade level
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

int count_letters(string text)
{

    // Initalise counter variable
    int letters = 0;
    // Loop over every character
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }
    // Return the number of letters in text
    return letters;
}

int count_words(string text)
{
    int words = 0;
    // Count number of spaces
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == ' ')
        {
            words++;
        }
    }
    // Number of spaces + 1 = number of words in sentnece
    words = words + 1;
    // Return the number of words in text
    return words;
}

int count_sentences(string text)
{

    int sentences = 0;

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        // Count punctuations
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }
    // Return the number of sentences in text
    return sentences;
}